### PR TITLE
HDFS-14541. When evictableMmapped or evictable size is zero, do not throw NoSuchElementException

### PR DESCRIPTION
**Pull request by leosunli**
…lementException

Signed-off-by: sunlisheng <sunlisheng@xiaomi.com>

### CODE REVIEW

File: hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/shortcircuit/ShortCircuitCache.java [hunk 1]
```diff
@@ -540,16 +538,18 @@ private void trimEvictionMaps() {
...
+      // throw NoSuchElementException
+      // if evictableMmappedSize == 0, evictableSize > 0 evictable do not throw
+      // NoSuchElementException
+      if (evictableSize == 0) {
+        replica = (ShortCircuitReplica) evictableMmapped
+            .get(evictableMmapped.firstKey());
+      } else {
+        replica = (ShortCircuitReplica) evictable.get(evictable.firstKey());
       }
+
```
Comment (hadoop-yetus) [inline hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/shortcircuit/ShortCircuitCache.java]: whitespace:end of line

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 30 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| -1 | test4tests | 0 | The patch doesn't appear to include any new or modified tests.  Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1390 | trunk passed |
| +1 | compile | 56 | trunk passed |
| +1 | checkstyle | 26 | trunk passed |
| +1 | mvnsite | 55 | trunk passed |
| +1 | shadedclient | 886 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 32 | trunk passed |
| 0 | spotbugs | 165 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 161 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 53 | the patch passed |
| +1 | compile | 48 | the patch passed |
| +1 | javac | 48 | the patch passed |
| +1 | checkstyle | 19 | the patch passed |
| +1 | mvnsite | 51 | the patch passed |
| -1 | whitespace | 0 | The patch has 1 line(s) that end in whitespace. Use git apply --whitespace=fix <<patch_file>>. Refer https://git-scm.com/docs/git-apply |
| +1 | shadedclient | 846 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 32 | the patch passed |
| +1 | findbugs | 167 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 63 | hadoop-hdfs-client in the patch failed. |
| +1 | asflicense | 34 | The patch does not generate ASF License warnings. |
| | | 4082 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.web.TestTokenAspect |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-977/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/977 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 6d73f1cadc93 4.4.0-143-generic #169~14.04.2-Ubuntu SMP Wed Feb 13 15:00:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 304a47e |
| Default Java | 1.8.0_212 |
| whitespace | https://builds.apache.org/job/hadoop-multibranch/job/PR-977/1/artifact/out/whitespace-eol.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-977/1/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs-client.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-977/1/testReport/ |
| Max. process+thread count | 307 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs-client U: hadoop-hdfs-project/hadoop-hdfs-client |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-977/1/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 75 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| -1 | test4tests | 0 | The patch doesn't appear to include any new or modified tests.  Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1120 | trunk passed |
| +1 | compile | 43 | trunk passed |
| +1 | checkstyle | 22 | trunk passed |
| +1 | mvnsite | 45 | trunk passed |
| +1 | shadedclient | 756 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 29 | trunk passed |
| 0 | spotbugs | 127 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 126 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 39 | the patch passed |
| +1 | compile | 37 | the patch passed |
| +1 | javac | 37 | the patch passed |
| +1 | checkstyle | 16 | the patch passed |
| +1 | mvnsite | 40 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 807 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 30 | the patch passed |
| +1 | findbugs | 142 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 117 | hadoop-hdfs-client in the patch passed. |
| +1 | asflicense | 31 | The patch does not generate ASF License warnings. |
| | | 3586 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=18.09.5 Server=18.09.5 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-977/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/977 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 4b4a9d84104c 4.15.0-48-generic #51-Ubuntu SMP Wed Apr 3 08:28:49 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 335c1c9 |
| Default Java | 1.8.0_212 |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-977/2/testReport/ |
| Max. process+thread count | 306 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs-client U: hadoop-hdfs-project/hadoop-hdfs-client |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-977/2/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 30 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| -1 | test4tests | 0 | The patch doesn't appear to include any new or modified tests.  Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1037 | trunk passed |
| +1 | compile | 44 | trunk passed |
| +1 | checkstyle | 24 | trunk passed |
| +1 | mvnsite | 48 | trunk passed |
| +1 | shadedclient | 735 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 31 | trunk passed |
| 0 | spotbugs | 124 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 122 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 41 | the patch passed |
| +1 | compile | 37 | the patch passed |
| +1 | javac | 37 | the patch passed |
| +1 | checkstyle | 16 | the patch passed |
| +1 | mvnsite | 42 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 742 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 27 | the patch passed |
| +1 | findbugs | 133 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 129 | hadoop-hdfs-client in the patch passed. |
| +1 | asflicense | 45 | The patch does not generate ASF License warnings. |
| | | 3385 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-977/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/977 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 181fc0d8940b 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 335c1c9 |
| Default Java | 1.8.0_212 |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-977/3/testReport/ |
| Max. process+thread count | 444 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs-client U: hadoop-hdfs-project/hadoop-hdfs-client |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-977/3/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

File: hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/shortcircuit/ShortCircuitCache.java [hunk 2]
```diff
@@ -533,23 +523,15 @@ private void trimEvictionMaps() {
...
-      try {
-        if (evictableSize == 0) {
-          replica = (ShortCircuitReplica)evictableMmapped.get(evictableMmapped
-              .firstKey());
-        } else {
-          replica = (ShortCircuitReplica)evictable.get(evictable.firstKey());
-        }
-      } catch (NoSuchElementException e) {
-        break;
+      if (evictable.size() == 0) {
```
Comment (goiri) [inline hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/shortcircuit/ShortCircuitCache.java]: Use isEmpty() here for clarity?

Comment (leosunli) [reply hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/shortcircuit/ShortCircuitCache.java]: @goiri Thanks for your comments . I have updated this patch.  Could you have time to  help review this patch? Thank you.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 34 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| -1 | test4tests | 0 | The patch doesn't appear to include any new or modified tests.  Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1171 | trunk passed |
| +1 | compile | 42 | trunk passed |
| +1 | checkstyle | 19 | trunk passed |
| +1 | mvnsite | 49 | trunk passed |
| +1 | shadedclient | 714 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 25 | trunk passed |
| 0 | spotbugs | 128 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 126 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 41 | the patch passed |
| +1 | compile | 38 | the patch passed |
| +1 | javac | 38 | the patch passed |
| +1 | checkstyle | 16 | the patch passed |
| +1 | mvnsite | 39 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 760 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 25 | the patch passed |
| +1 | findbugs | 136 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 117 | hadoop-hdfs-client in the patch passed. |
| +1 | asflicense | 25 | The patch does not generate ASF License warnings. |
| | | 3467 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-977/4/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/977 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux fd467a36ce5a 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / b28ddb2 |
| Default Java | 1.8.0_212 |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-977/4/testReport/ |
| Max. process+thread count | 447 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs-client U: hadoop-hdfs-project/hadoop-hdfs-client |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-977/4/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 44 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| -1 | test4tests | 0 | The patch doesn't appear to include any new or modified tests.  Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1227 | trunk passed |
| +1 | compile | 41 | trunk passed |
| +1 | checkstyle | 22 | trunk passed |
| +1 | mvnsite | 44 | trunk passed |
| +1 | shadedclient | 797 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 30 | trunk passed |
| 0 | spotbugs | 136 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 131 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 45 | the patch passed |
| +1 | compile | 38 | the patch passed |
| +1 | javac | 38 | the patch passed |
| +1 | checkstyle | 18 | the patch passed |
| +1 | mvnsite | 41 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 840 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 28 | the patch passed |
| +1 | findbugs | 138 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 109 | hadoop-hdfs-client in the patch passed. |
| +1 | asflicense | 28 | The patch does not generate ASF License warnings. |
| | | 3725 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-977/5/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/977 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux b2a98ebf104b 4.4.0-143-generic #169~14.04.2-Ubuntu SMP Wed Feb 13 15:00:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / b28ddb2 |
| Default Java | 1.8.0_212 |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-977/5/testReport/ |
| Max. process+thread count | 336 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs-client U: hadoop-hdfs-project/hadoop-hdfs-client |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-977/5/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by goiri** [comment]
LGTM

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 0 | Docker mode activated. |
| -1 | patch | 12 | https://github.com/apache/hadoop/pull/977 does not apply to trunk. Rebase required? Wrong Branch? See https://wiki.apache.org/hadoop/HowToContribute for help. |


| Subsystem | Report/Notes |
|----------:|:-------------|
| GITHUB PR | https://github.com/apache/hadoop/pull/977 |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-977/6/console |
| versions | git=2.7.4 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by openinx** [comment]
@goiri ,  I think we need to cherry-pick this feature to all branches, such as branch-2.9 etc. Thanks.

**Comment by goiri** [comment]
Does github allow cherry-picking? git cherry-pick -x requires -m which is not clear to me.

**Comment by openinx** [comment]
I saw you have done the cherry-pick, Thanks @goiri for your work :-)

**Comment by goiri** [comment]
For full disclosure, I messed up the merge of the PR and I had to do the cherry-pick by hand through the patch.
Next time I should do a proper squash and merge and hopefully cherry-pick will go clean.

**Pull request merged**